src/runtime/hexagon_remote/halide_remote.cpp

```diff
@@ -147,72 +148,163 @@ handle_t halide_hexagon_remote_get_symbol(handle_t module_ptr, const char* name,
 
-        HAP_power_request_t request;
+    power_info.type = HAP_power_get_max_mips;
+    int retval = HAP_power_get(NULL, &power_info);
+    if (0 != retval) {
+        log_printf("HAP_power_get(HAP_power_get_max_mips) failed (%d)\n", retval);
+        return -1;
+    }
+    unsigned int max_mips = power_info.max_mips;
 
-        request.type = HAP_power_set_apptype;
-        request.apptype = HAP_POWER_COMPUTE_CLIENT_CLASS;
-        retval = HAP_power_set(NULL, &request);
-        if (0 != retval) {
-            log_printf("HAP_power_set(HAP_power_set_apptype) failed (%d)\n", retval);
-            return -1;
-        }
+    power_info.type = HAP_power_get_max_bus_bw;
+    retval = HAP_power_get(NULL, &power_info);
+    if (0 != retval) {
+        log_printf("HAP_power_get(HAP_power_get_max_bus_bw) failed (%d)\n", retval);
+        return -1;
+    }
+    uint64 max_bus_bw = power_info.max_bus_bw;
+
+    // The above API under-reports the max bus bw. If we use it as
+    // reported, performance is bad. Experimentally, this only
+    // needs to be ~10, but since it's wrong, we might as well
+    // have a safety factor...
+    max_bus_bw *= 1000;
+
+    // Since max_bus_bw is bad, might as well make sure max_mips
+    // isn't bad too.
+    max_mips *= 1000;
+
+    perf->set_mips       = TRUE;
+    perf->mipsPerThread  = max_mips;
+    perf->mipsTotal      = max_mips;
+    perf->set_bus_bw     = TRUE;
+    perf->bwBytePerSec   = max_bus_bw;
+    perf->busbwUsagePercentage = 100;
+    perf->set_latency    = TRUE;
+    perf->latency        = 1;
 
-        request.type = HAP_power_set_HVX;
-        request.hvx.power_up = TRUE;
-        retval = HAP_power_set(NULL, &request);
-        if (0 != retval) {
-            log_printf("HAP_power_set(HAP_power_set_HVX) failed (%d)\n", retval);
-            return -1;
-        }
+    return 0;
+}
 
-        request.type = HAP_power_set_mips_bw;
-        request.mips_bw.set_mips = TRUE;
-        request.mips_bw.mipsPerThread = max_mips;
-        request.mips_bw.mipsTotal = max_mips;
-        request.mips_bw.set_bus_bw = TRUE;
-        request.mips_bw.bwBytePerSec = max_bus_bw;
-        request.mips_bw.busbwUsagePercentage = 100;
-        request.mips_bw.set_latency = TRUE;
-        request.mips_bw.latency = 1;
-        retval = HAP_power_set(NULL, &request);
-        if (0 != retval) {
-            log_printf("HAP_power_set(HAP_power_set_mips_bw) failed (%d)\n", retval);
-            return -1;
+int halide_hexagon_remote_power_hvx_on_perf(
+    int set_mips,
+    int mipsPerThread,
+    int mipsTotal,
+    int set_bus_bw,
+    int bwBytePerSec_u32,
+    int bwBytePerSec_l32,
+    int busbwUsagePercentage,
+    int set_latency,
+    int latency)
+{
+    int result = 0;
+    if (power_ref_count == 0) {
+        halide_hvx_power_perf_t perf;
+        uint64_t bwBytePerSec = 0;
+
+        perf.set_mips       = set_mips;
+        perf.mipsPerThread  = *((unsigned int *) &mipsPerThread);
```

Comment:
These are reinterpret casts since the IDL interface forces all parameters to be passed as ints.  This treats the full 32-bits as an unsigned int (which is what was actually passed through).